# <span style="color:teal;">CIS 211 Project 7:  SQLite</span>

##### Due 11:00 P.M. Tuesday Mar 7

This project will give you some experience writing database queries in SQL and connecting to a SQLite database from Python.

### <span style="color:teal">Setup</span>

Download the Sakila database (the file named `sakila211.db`) and move it to the same folder as this IPython notebook.

### <span style="color:teal">Executing Queries in an IPython Notebook</span>

To solve the problems in the first two sections of this project you need to write a SQL query based on the tables in the Sakila211 database.  

To execute a query from an IPython Notebook write the query in quotes as part of a shell command:
```
! sqlite3 sakila211.db "..."
```

For example, to answer the question "What are the names and ratings of the films" your code cell would have
```
! sqlite3 sakila211.db "select title, rating from film"
```

Some of the queries will produce a lot of output.  You can add a `LIMIT` clause to any query if you want to reduce the size of the output.

**Note:** &nbsp; Your grade will not depend on the output format.  If want to have a nicer looking output that includes column headers create a file named `.sqliterc` in your home folder and add these two lines:
```
.header on
.mode column
```

## <span style="color:teal">Part 1:  &nbsp; Queries</span>

The following questions can all be answered using records from a single table.

(1) What are the first and last names of all the customers?

In [1]:
! sqlite3 sakila211.db "SELECT first_name, last_name FROM customer"

Mary|Smith
Patricia|Johnson
Linda|Williams
Barbara|Jones
Elizabeth|Brown
Jennifer|Davis
Maria|Miller
Susan|Wilson
Margaret|Moore
Dorothy|Taylor
Lisa|Anderson
Nancy|Thomas
Karen|Jackson
Betty|White
Helen|Harris
Sandra|Martin
Donna|Thompson
Carol|Garcia
Ruth|Martinez
Sharon|Robinson
Michelle|Clark
Laura|Rodriguez
Sarah|Lewis
Kimberly|Lee
Deborah|Walker
Jessica|Hall
Shirley|Allen
Cynthia|Young
Angela|Hernandez
Melissa|King
Brenda|Wright
Amy|Lopez
Anna|Hill
Rebecca|Scott
Virginia|Green
Kathleen|Adams
Pamela|Baker
Martha|Gonzalez
Debra|Nelson
Amanda|Carter
Stephanie|Mitchell
Carolyn|Perez
Christine|Roberts
Marie|Turner
Janet|Phillips
Catherine|Campbell
Frances|Parker
Ann|Evans
Joyce|Edwards
Diane|Collins
Alice|Stewart
Julie|Sanchez
Heather|Morris
Teresa|Rogers
Doris|Reed
Gloria|Cook
Evelyn|Morgan
Jean|Bell
Cheryl|Murphy
Mildred|Bailey
Katherine|Rivera
Joan|Cooper
Ashley|Richardson
Judith|Cox
Rose|Howard
Janice|Ward
Kelly|Torres
Nicole|Peterson
Judy|Gray
Christina|Ramirez
Kathy|James
Theresa

(2) What are the e-mail addresses of customers whose first name is “Willard”?

In [43]:
! sqlite3 sakila211.db "SELECT email FROM customer WHERE first_name = 'Willard'"

willard.lumpkin@sakilacustomer.org


(3) How many customers are there?

In [44]:
! sqlite3 sakila211.db "SELECT count(first_name) FROM customer"

599


(4) How many customers shop at store number 1?

In [45]:
! sqlite3 sakila211.db "SELECT count(store_id) FROM customer WHERE store_id = 1"

326


(5) How much does it cost to rent the film named “Virtual Spoilers”?

In [46]:
! sqlite3 sakila211.db "SELECT rental_rate FROM film WHERE title = 'Virtual Spoilers'";

4.99


(6) Do any films have the word “Princess” in the title?

In [47]:
! sqlite3 sakila211.db "SELECT EXISTS(SELECT title FROM film WHERE title LIKE '%Princess%')"

1


(7) What are the titles of the films that are longer than 180 minutes?

In [48]:
! sqlite3 sakila211.db "SELECT title FROM film WHERE length > 180"

Analyze Hoosiers
Baked Cleopatra
Catch Amistad
Chicago North
Conspiracy Spirit
Control Anthem
Crystal Breaking
Darn Forrester
Frontier Cabin
Gangs Pride
Haunting Pianist
Home Pity
Hotel Happiness
Intrigue Worst
Jacket Frisco
King Evolution
Lawless Vision
Love Suicides
Monsoon Cause
Moonwalker Fool
Muscle Bright
Pond Seattle
Records Zorro
Reds Pocus
Runaway Tenenbaums
Saturn Name
Scalawag Duck
Searchers Wait
Smoochy Control
Soldiers Evolution
Sons Interview
Sorority Queen
Star Operation
Sweet Brotherhood
Theory Mermaid
Wife Turn
Wild Apollo
Worst Banger
Young Language


(8) How many films have a “G” rating and are less than 60 minutes long?

In [49]:
! sqlite3 sakila211.db "SELECT count(title) FROM film WHERE rating = 'G' AND length < 60"

21


(9) What is the maximum replacement cost for any film?

In [50]:
! sqlite3 sakila211.db "SELECT MAX(replacement_cost) FROM film"

29.99


(10) Print a table that lists the different types of ratings and the number of films that have that rating.

In [51]:
! sqlite3 sakila211.db "SELECT rating, count(*)  FROM film GROUP BY rating"

G|178
NC-17|210
PG|194
PG-13|223
R|195


## <span style="color:teal">Part 2: &nbsp; Joins</span>

The following queries all require a join of two or more tables. As a hint for how to create the query the table names are shown next to each a question.

(11) What are the film IDs of movies starring Jude Cruise?  &nbsp; &nbsp; [`actor`, `film_actor`]

In [52]:
! sqlite3 sakila211.db "SELECT film_id FROM film_actor JOIN actor USING (actor_id) WHERE actor_id = 57"

16
34
101
114
122
134
144
153
192
213
258
267
317
340
393
437
447
502
592
605
637
685
707
714
717
737
767
852
891
918


(12) What language was “Ace Goldfinger” filmed in?  &nbsp; &nbsp; [`film`, `language`]

In [54]:
! sqlite3 sakila211.db "SELECT name FROM language JOIN film USING (language_id) WHERE title = 'Ace Goldfinger'"

Japanese


(13) How many actors starred in the film named “Scalawag Duck”? &nbsp; &nbsp; [`film`, `film_actor`]

In [55]:
! sqlite3 sakila211.db "SELECT count(actor_id) FROM film_actor JOIN film USING(film_id) WHERE title = 'Scalawag Duc'"

0


(14) Produce a table that shows how many films each customer has rented.  &nbsp; &nbsp; [`customer`, `rental`]

In [56]:
! sqlite3 sakila211.db "SELECT first_name, count(customer_id) AS 'Rented amt' FROM rental JOIN customer USING(customer_id) GROUP BY first_name"

Aaron|24
Adam|22
Adrian|18
Agnes|23
Alan|26
Albert|22
Alberto|20
Alex|33
Alexander|36
Alfred|26
Alfredo|20
Alice|33
Alicia|20
Allan|18
Allen|20
Allison|25
Alma|35
Alvin|29
Amanda|27
Amber|27
Amy|29
Ana|33
Andre|28
Andrea|22
Andrew|27
Andy|24
Angel|32
Angela|35
Anita|15
Ann|17
Anna|20
Anne|23
Annette|22
Annie|18
Anthony|20
Antonio|16
April|25
Arlene|26
Armando|21
Arnold|33
Arthur|32
Ashley|25
Audrey|29
Austin|19
Barbara|22
Barry|32
Beatrice|26
Becky|28
Ben|26
Benjamin|23
Bernard|21
Bernice|33
Bertha|25
Bessie|28
Beth|24
Betty|27
Beverly|23
Bill|26
Billie|26
Billy|35
Bob|24
Bobbie|24
Bobby|35
Bonnie|21
Brad|25
Bradley|27
Brandon|36
Brandy|28
Brenda|26
Brent|22
Brett|34
Brian|12
Brittany|28
Bruce|23
Bryan|28
Byron|29
Calvin|23
Carl|23
Carla|26
Carlos|23
Carmen|25
Carol|22
Carole|30
Caroline|15
Carolyn|28
Carrie|34
Casey|34
Cassandra|28
Catherine|34
Cathy|27
Cecil|25
Chad|25
Charlene|27
Charles|32
Charlie|25
Charlotte|24
Cheryl|27
Chester|24
Chris|21
Christian|23
Christina|18
Christine|22


(15) How many films did the customer named Smith rent?  &nbsp; &nbsp; [`customer`, `rental`]

In [57]:
! sqlite3 sakila211.db "SELECT first_name, last_name, count(customer_id) AS 'Rented amt' FROM rental JOIN customer USING(customer_id) WHERE first_name LIKE '%smith%' OR last_name LIKE '%smith%'"

Mary|Smith|32


(16) Which actors starred in “Splash Gump”?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [58]:
! sqlite3 sakila211.db "SELECT first_name, last_name FROM actor JOIN film_actor USING(actor_id) JOIN film USING(film_id) WHERE title = 'Splash Gump'"

Penelope|Guiness
Jennifer|Davis
Uma|Wood
Cameron|Streep
Christian|Neeson
Ray|Johansson
Gene|Willis
Morgan|Williams
Humphrey|Willis
Al|Garland


(17) Which films has Penelope Guiness starred in?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [59]:
! sqlite3 sakila211.db "SELECT title FROM film JOIN film_actor USING(film_id) JOIN actor USING(actor_id) WHERE first_name = 'Penelope' AND last_name = 'Guiness'"

Academy Dinosaur
Anaconda Confessions
Angels Life
Bulworth Commandments
Cheaper Clyde
Color Philadelphia
Elephant Trojan
Gleaming Jawbreaker
Human Graffiti
King Evolution
Lady Stage
Language Cowboy
Mulholland Beast
Oklahoma Jumanji
Rules Human
Splash Gump
Vertigo Northwest
Westward Seabiscuit
Wizard Coldblooded


(18) What are the names of the films rented by the customer named Knight?  &nbsp; &nbsp; [`customer`, `rental`, `inventory`, `film`]

In [60]:
! sqlite3 sakila211.db "SELECT title FROM film JOIN inventory USING(film_id) JOIN rental USING(inventory_id) JOIN customer USING(customer_id) WHERE first_name LIKE '%knight%' OR last_name LIKE '%knight%'"

Jericho Mulan
Frost Head
Clerks Angels
Wife Turn
Expecations Natural
Lawless Vision
Soup Wisdom
Heartbreakers Bright
Bikini Borrowers
Attacks Hate
Cupboard Sinners
Princess Giant
Cyclone Family
Platoon Instinct
Knock Warlock
Excitement Eve
Right Cranes
Fatal Haunted
Falcon Volume
Tomorrow Hustler
Half Outfield
Horror Reign
Borrowers Bedazzled


## <span style="color:teal">Part 3: &nbsp; Monthly Report</span>

For this part of the project you will write a report generator that will print a summary of monthly activity for a customer.  The report should show a list of all the films rented during the month along with a total amount due.
This example shows the activity for the customer named Leo Ebert for the month of June, 2005.

<pre>
--- Sakila DVD Rentals ---

Monthly Report for Leo Ebert 

Lonely Elephant                 06/16/2005    $2.99
                   ** late fee  06/24/2005    $2.99
Enemy Odds                      06/18/2005    $4.99
                   ** late fee  06/27/2005    $4.99
Song Hedwig                     06/20/2005    $0.99
                   ** late fee  06/26/2005    $0.99
Wash Heavenly                   06/21/2005    $4.99
Hurricane Affair                06/21/2005    $2.99

Monthly Total:  $25.92
</pre>

To compute the total, the report generator needs to add up all the rental fees, and for any film returned late, add an additional rental fee as the late charge. In the example, Leo rented *Lonely Elephant* on June 16 but didn’t return it until June 24. It was a 3-day rental, so he was charged an additional \$2.99 (it was 5 days late, so we could have charged him \$4.98 in late fees, but we’ll keep it simple and just charge the basic fee as the late fee).

We've broken the problem into separate parts so they can be implemented and tested separately.  Your job is to fill in the code cells so each part passes the auto-grader tests.

**Note:** &nbsp; There is no documentation requirement for this project.

### <span style="color:teal">Database Connection</span>

To simplify the code for this program you can define the database connection as a global variable.  The following code cell imports the libraries you need and defines the database connection.

In [61]:
from sqlite3 import connect

db = connect('sakila211.db')

### <span style="color:teal">Dates</span>

Dates in a SQLite database are stored as strings.  The function below uses a Python library named `datetime` to parse a date string and return a string that can be printed as part of the report.  Read the online documentation for the `datetime` module if you want to learn more about date formats and functions that parse and format dates.

**Note:** &nbsp; The format string is a global variable -- you can use this same format in your function that checks to see if a CD is overdue.

In [62]:
from datetime import datetime

date_format = "%Y-%m-%d %H:%M:%S.%f"

def mmddyyyy(date):
    d = datetime.strptime(date, date_format)
    return '{:02d}/{:02d}/{:02d}'.format(d.month, d.day, d.year)

In [63]:
assert mmddyyyy('2005-02-26 20:59:35.000') == '02/26/2005'

### <span style="color:teal">Customer Name</span>

Fill in the definition of the `customer_name` function.  It will be passed a string with a last name, and it should return a list of tuples of the form `(first, last)` with the first and last names of all customers with the same last name.  If there are no customers with that name the list should be empty.

In [64]:
def customer_names(s):
    sub = 'SELECT first_name, last_name FROM customer WHERE last_name LIKE ?'
    result = db.execute(sub, (s, ))
    return list(result)

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [65]:
assert customer_names('ebert') == [('Leo', 'Ebert')]

In [66]:
assert customer_names('smith') == [('Mary', 'Smith')]

In [67]:
assert customer_names('turing') == []

### <span style="color:teal">Film List</span>

Fill in the body of the `film_list` function so it returns a list of all CDs rented by a customer during a specified month. 
The parameters that will be passed to the function are the customer's last name (a string) and the month and year (both integers).  Return a list of tuples that have the following values, in the order shown:
* film title
* rental rate (how much it costs to rent the CD)
* rental date (the date the CD was rented)
* return date (the date the CD was returned)
* rental duration (the number of days the CD can be out before a late fee is charged) 

All of the information you need can be retrieved with a single query that joins the rental, inventory, film, and customer tables. Note that this is the same query used to answer Question 19 above, except with different column names.

**Hint:** &nbsp; The dates in the database are strings.  Think about pattern matching when you write the WHERE clause that specifies the date range.


In [68]:
def film_list(name, month, year):
    sub = 'SELECT title, rental_rate, rental_date, return_date, rental_duration FROM film JOIN inventory USING(film_id) JOIN rental USING(inventory_id) JOIN customer USING(customer_id) WHERE last_name LIKE ? AND rental_date LIKE ?'
    month_year = '%{}-0{}%'.format(year, month)
    result = db.execute(sub, (name, month_year))
    return list(result)

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [69]:
leo = film_list('Ebert', 6, 2005)
assert len(leo) == 5
assert leo[0] == (('Lonely Elephant', 2.99, '2005-06-16 20:59:35.000', '2005-06-24 15:31:35.000', 3))

In [70]:
mary = film_list('Smith', 8, 2005)
assert len(mary) == 11
assert mary[0] == ('Adaptation Holes', 2.99, '2005-08-01 08:51:04.000', '2005-08-10 12:12:04.000', 7)

### <span style="color:teal">Overdue CDs</span>

To decide whether a CD was returned late you need to subtract the rental date from the return date and see if the difference is greater than the rental duration. The easiest way to do this is to use Python’s `datetime` library.

Using a call to `strptime` (see the body of the `mmddyyyy` function defined above) create `datetime` objects for the rental date and the return date. Then subtract the rental date from the return date to get a `timedelta` object:
```
    diff = returned - rented
```
Now you can get attributes of the object named `diff` to tell you how far apart the dates were, for example `diff.days` will tell you the time difference in days.

Fill in the body of the `overdue` function.  The parameters are a rental date, a return date, and the rental duration.  Return TRUE if the CD was not returned or if the number of days it was checked out is greater than the rental duration.

In [71]:
def overdue(rent_date, return_date, duration):
    rented = datetime.strptime(rent_date, date_format)
    if return_date != None:
        returned = datetime.strptime(return_date, date_format)
    else:
        return False

    diff = returned - rented
    actual_diff = ((abs(diff.days)*86400) - diff.seconds) / 86400
    
    return True if actual_diff > duration else False

In [72]:
assert overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:35.000', 2)

In [73]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-17 20:59:35.000', 2)

In [74]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:34.000', 2)

### <span style="color:teal">Report Generator</span>

Fill in the body of the function named `sakila_report` so it prints a monthly report.  The arguments passed to the function will be a customer's last name (a string) and a month and year (both integers).

If the customer is not in the database, or if there are two customers with the same last name, return None.  Otherwise print a list of films and late fees and the total amount due.

You can format the report however you would like, as long as the following items appear somewhere in the output:
* every fee should be printed with a dollar sign and an amount, for example `$2.99` or `25.92`
* the total amound due should be printed on a line that includes the word `"Total"` followed by a dollar amount on the same line

In [75]:
def sakila_report(customer, month, year):
    sub = 'SELECT first_name FROM customer WHERE last_name = ?'
    try:
        result = db.execute(sub, (customer, ))
    except:
        return None
    
    listnum = len(list(result))
    if  listnum > 1 or listnum <= 0:
        return None
    
    clist = film_list(customer, month, year)
    first = list(db.execute('SELECT first_name FROM customer WHERE last_name = ?', (customer, )))[0][0]
    total = 0
    
    print('------------- *** SAKILA DVD Rentals *** -------------')
    print('\nMonthly Report for {} {} \n'.format(first, customer))
    template = "{0:22}{1:12}| {2} -> ${3}"
    for rental in clist: 
        od = overdue(rental[2], rental[3], rental[4])
        total = total + rental[1]
        print(template.format(rental[0], '', mmddyyyy(rental[2]), rental[1]))
        if od:
            print(template.format('', '** late fee', mmddyyyy(rental[3]), rental[1]))
            total = total + rental[1]
    
    print('\nMonthly Total: ${:.4}'.format(total))

In [76]:
sakila_report('Smith', 8, 2005)

------------- *** SAKILA DVD Rentals *** -------------

Monthly Report for Mary Smith 

Adaptation Holes                  | 08/01/2005 -> $2.99
                      ** late fee | 08/10/2005 -> $2.99
Racer Egg                         | 08/02/2005 -> $2.99
                      ** late fee | 08/10/2005 -> $2.99
Finding Anaconda                  | 08/02/2005 -> $0.99
Minds Truman                      | 08/17/2005 -> $4.99
Dalmations Sweden                 | 08/18/2005 -> $0.99
Patient Sister                    | 08/19/2005 -> $0.99
Jumanji Blade                     | 08/19/2005 -> $2.99
Unforgiven Zoolander              | 08/21/2005 -> $0.99
Fireball Philadelphia             | 08/22/2005 -> $0.99
                      ** late fee | 08/27/2005 -> $0.99
Fireball Philadelphia             | 08/22/2005 -> $0.99
                      ** late fee | 08/28/2005 -> $0.99
Bikini Borrowers                  | 08/22/2005 -> $4.99

Monthly Total: $32.85


##### <span style="color:red">Tests</span>

You can use code cells in this section (and add more if you wish) to test your functions.

##### <span style="color:red">Autograder Tests</span>

The tests for the `sakila_report` function will call the function and while using the IPython `capture` command to save everything the function prints.  We'll then use a function that counts the number of times a dollar amount (from either a rental fee or a late fee) occurs and make sure the total amount is the expected value.

**Do not edit, copy, or delete these cells.**

In [77]:
import re

def check_report(text, expected_count, expected_total):
    nums = []
    for line in text.split('\n'):
        has_total = re.search(r'Total.*(\$\d+\.\d\d)', line)
        if has_total:
            assert has_total.group(1) == expected_total
        else:
            for s in re.findall(r'\$\d+\.\d\d', line):
                nums.append(s)
    assert(len(nums) == expected_count)
    total = 0.0
    for s in nums:
        total += float(s[1:])
    assert(round(total,2)) == round(float(expected_total[1:]), 2)
    return True

In [78]:
%%capture res
sakila_report('Ebert', 6, 2005)

In [79]:
assert check_report(res.stdout, 8, '$25.92')

In [80]:
%%capture res
sakila_report('Smith', 8, 2005)

In [81]:
assert check_report(res.stdout, 15, '$32.85')

In [82]:
%%capture res
sakila_report('Turing', 6, 2005)

In [83]:
assert len(res.stdout) == 0

In [84]:
sakila_report('Ebert', 6, 2005)

------------- *** SAKILA DVD Rentals *** -------------

Monthly Report for Leo Ebert 

Lonely Elephant                   | 06/16/2005 -> $2.99
                      ** late fee | 06/24/2005 -> $2.99
Enemy Odds                        | 06/18/2005 -> $4.99
                      ** late fee | 06/27/2005 -> $4.99
Song Hedwig                       | 06/20/2005 -> $0.99
                      ** late fee | 06/26/2005 -> $0.99
Wash Heavenly                     | 06/21/2005 -> $4.99
Hurricane Affair                  | 06/21/2005 -> $2.99

Monthly Total: $25.92
